src/libs/logging/src/file_rotate.rs

```diff
@@ -0,0 +1,276 @@
+            }
+            Err(e) if e.kind() == io::ErrorKind::NotFound => {}
+            Err(e) => return Err(e),
+        }
+        if let Some(parent) = p.parent() {
+            if p.has_root() || !parent.as_os_str().is_empty() {
+                let md = parent.metadata()?;
+                if !md.is_dir() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("'{}' is not a directory", parent.to_string_lossy()),
+                    ));
+                }
+            }
+        }
+
+        Ok(FileRotator {
+            path: p.to_path_buf(),
+            file: None,
+            ignore_errors: false,
+            rotate_size: DEFAULT_LOG_FILE_SIZE_TO_ROTATE,
+            rotate_keep: DEFAULT_HISTORY_LOG_FILES,
+            truncate: false,
+            written_size: 0,
+        })
+    }
+
+    /// Use "truncate" or "append" mode when opening the log file.
+    pub fn truncate_mode(&mut self, truncate: bool) -> &mut Self {
+        self.truncate = truncate;
+        self
+    }
+
+    /// Set the threshold size to rotate log files.
+    pub fn rotate_threshold(&mut self, size: u64) -> &mut Self {
+        self.rotate_size = size;
+        self
+    }
+
+    /// Set number of rotated log files to keep.
+    pub fn rotate_count(&mut self, count: usize) -> &mut Self {
+        self.rotate_keep = count;
+        self
+    }
+
+    /// Ignore all errors and try best effort to log messages but without guarantee.
+    pub fn ignore_errors(&mut self, ignore_errors: bool) -> &mut Self {
+        self.ignore_errors = ignore_errors;
+        self
+    }
+
+    /// Open the log file if
+    /// - it hasn't been opened yet.
+    /// - current log file has been rotated and needs to open a new log file.
+    fn reopen_if_needed(&mut self) -> Result<()> {
+        if self.file.is_none() || !self.path.exists() {
+            let file = OpenOptions::new()
+                .create(true)
+                .write(true)
+                .truncate(self.truncate)
+                .append(!self.truncate)
+                .open(&self.path)?;
+            match file.metadata() {
+                Ok(md) => self.written_size = md.len(),
+                Err(e) => {
+                    if self.ignore_errors {
+                        // Pretend as an empty file.
+                        // It's better to permit over-sized log file instead of disabling rotation.
+                        self.written_size = 0;
+                    } else {
+                        return Err(e);
+                    }
+                }
+            }
+            self.file = Some(LineWriter::new(file));
+        }
+
+        Ok(())
+    }
+
+    /// Try to rotate log files.
+    ///
+    /// When failed to rotate the log files, we choose to ignore the error instead of possibly
+    /// panicking the whole program. This may cause over-sized log files, but that should be easy
+    /// to recover.
```

Comment:
Please can you add an explicit test to ensure that if the rotation fails, the original log file continues to be updated.

Reply:
It's not ease to trigger a failure for rotate. Actually we need to add special code for test to support this.

Reply:
done

Reply:
I was imaginging a test that would only run as non-`root` which would:

- Create a rotator as the non-`root` user with the file specified as `log.txt`.
- Create `log.txt.1` as an unwritable file (owned by `root` by calling `sudo`).
- Write to the rotator and check it works.
- Attempt to rotate (which should fail).
- Write again and check that the original file contains both writes.